src/AzureExtension/DataManager/CacheManager.cs

```diff
@@ -0,0 +1,323 @@
+                            LastUpdated = DateTime.UtcNow;
+                        }
+                    }
+
+                    _log.Information($"Accounts updated: {e.Context.AccountsUpdated}  Skipped: {e.Context.AccountsSkipped}  Errors: {e.Context.Errors}  Elapsed: {e.Context.TimeElapsed}");
+                    if (e.Exception is not null)
+                    {
+                        _log.Warning($"First Error: {e.Exception.Message}");
+                    }
+
+                    SendUpdateEvent(this, CacheManagerUpdateKind.Updated);
+                    break;
+
+                case DataManagerUpdateKind.Cancel:
+                    lock (_stateLock)
+                    {
+                        UpdateInProgress = false;
+                    }
+
+                    _log.Information($"Operation was cancelled. Updated: {e.Context.AccountsUpdated}  Skipped: {e.Context.AccountsSkipped}  Errors: {e.Context.Errors}  Elapsed: {e.Context.TimeElapsed}");
+                    SendUpdateEvent(this, CacheManagerUpdateKind.Cancel);
+                    if (_pendingRefresh)
+                    {
+                        // If we were pending a refresh it was likely because a refresh caused this
+                        // cancellation, and there is a race between the update canncellation happening
+                        // and the subsequent update. If we get here it means that we possibly need to update.
+                        // If the Refresh call update executes first then this one will be ignored due
+                        // to update in progress.
+                        _ = Update(TimeSpan.MinValue);
+                    }
+
+                    break;
+
+                case DataManagerUpdateKind.Error:
+                    lock (_stateLock)
+                    {
+                        // Error condition means we dont' know what state we are in, but we want to clear everything so retries
+                        // can be allowed.
+                        UpdateInProgress = false;
+                        _pendingRefresh = false;
+                    }
+
+                    _log.Error(e.Exception, "Update failed.");
+                    SendUpdateEvent(this, CacheManagerUpdateKind.Error, e.Exception);
+                    break;
+            }
+        }
+    }
+
+    private void SendUpdateEvent(object? source, CacheManagerUpdateKind kind, Exception? ex = null)
+    {
+        if (OnUpdate != null)
+        {
+            _log.Debug($"Sending Update Event.  Kind: {kind}");
+            OnUpdate.Invoke(source, new CacheManagerUpdateEventArgs(kind, ex));
+        }
+    }
+
+    private void HandleDeveloperIdChange(IDeveloperIdProvider sender, IDeveloperId args)
+    {
+        if (DeveloperIdProvider.GetInstance().GetDeveloperIdState(args) == AuthenticationState.LoggedIn)
+        {
+            // New login means we trigger a new sync.
+            _log.Information("New developer account logged in, syncing data...");
```

Comment:
Does this trigger a new sync? Extra comment might be helpful since I just see a log :)

Reply:
Great catch, it did not. I have updated the code to and done some additional login/logout testing to verify we do a refresh whenever a developer id is logged in or out, and added additional comments to the behavior and why. I also wrapped this in a try-catch in case DeveloperIdProvider threw, and moved the state check to a switch statement for future proofing.